examples/PingPong.hs

```diff
@@ -37,38 +45,79 @@ instance Mockable SharedAtomic IO where
-    liftMockable (NewChannel) = Conc.newChan
-    liftMockable (ReadChannel channel) = Conc.readChan channel
-    liftMockable (WriteChannel channel t) = Conc.writeChan channel t
+    liftMockable (NewChannel) = STM.atomically Conc.newTChan
+    liftMockable (ReadChannel channel) = STM.atomically $ Conc.readTChan channel
+    liftMockable (TryReadChannel channel) = STM.atomically $ Conc.tryReadTChan channel
+    liftMockable (UnGetChannel channel t) = STM.atomically $ Conc.unGetTChan channel t
+    liftMockable (WriteChannel channel t) = STM.atomically $ Conc.writeTChan channel t
 
 instance Mockable Bracket IO where
     liftMockable (Bracket acquire release act) = Exception.bracket acquire release act
 
 instance Mockable Throw IO where
     liftMockable (Throw e) = Exception.throwIO e
 
-workers :: ( RandomGen g ) => NodeId -> g -> [NodeId] -> [Worker IO]
+instance Mockable Catch IO where
+    liftMockable (Catch action handler) = action `Exception.catch` handler
+
+-- Sending a message which encodes to "" is problematic!
+-- The receiver can't distinuish this from the case in which the sender sent
+-- nothing at all.
+-- So we give custom Ping and Pong types with non-generic Binary instances.
+--
+-- TBD should we fix this in network-transport? Maybe every chunk is prefixed
+-- by a byte giving its length? Wasteful I guess but maybe not a problem.
+
+data Ping = Ping
+deriving instance Show Ping
+instance Binary Ping where
+    put _ = putWord8 (fromIntegral 0)
+    get = do
+        w <- getWord8
+        if w == fromIntegral 0
+        then pure Ping
+        else fail "no parse ping"
+
+data Pong = Pong
+deriving instance Show Pong
+instance Binary Pong where
+    put _ = putWord8 (fromIntegral 1)
+    get = do
+        w <- getWord8
+        if w == fromIntegral 1
+        then pure Pong
+        else fail "no parse pong"
+
+workers :: NodeId -> StdGen -> [NodeId] -> [Worker IO]
 workers id gen peerIds = [pingWorker gen]
     where
-    pingWorker :: ( RandomGen g ) => g -> SendActions IO -> IO ()
+    pingWorker :: StdGen -> SendActions IO -> IO ()
     pingWorker gen sendActions = loop gen
         where
+        loop :: StdGen -> IO ()
         loop gen = do
             let (i, gen') = randomR (0,1000000) gen
-            putStrLn (show id ++ " is waiting for " ++ show i ++ "us before sending pings")
+            --putStrLn (show id ++ " is waiting for " ++ show i ++ "us before sending pings")
             Conc.threadDelay i
-            forM_ peerIds (\peerId -> sendTo sendActions peerId (fromString "ping") ())
+            let pong :: NodeId -> ConversationActions Void Pong IO -> IO ()
+                pong peerId cactions = do
+                    putStrLn (show id ++ " sent PING to " ++ show peerId)
+                    Pong <- recv cactions
+                    putStrLn (show id ++ " heard PONG from " ++ show peerId)
+            forM_ peerIds $ \peerId -> withConnectionTo sendActions peerId (fromString "ping") (pong peerId)
```

Comment:
Notice that the pinger never `send`s anything. The `MessageName` sent to the peer (`"ping"`) essentially fills the role of a send.